Sources/PointFree/GitHubApi.swift

```diff
@@ -39,43 +39,58 @@ func fetchAuthToken(forCode code: String) -> EitherIO<Prelude.Unit, GitHubAccess
       "Accept": "application/json"
   ]
 
-  return .init(
-    run: .init { callback in
-      session
-        .dataTask(with: request) { data, response, error in
-          callback(
-            data.flatMap { try? JSONDecoder().decode(GitHubAccessToken.self, from: $0) }
-              .map(Either.right)
-              ?? Either.left(unit)
-          )
-        }
-        .resume()
-    }
-  )
+  return (jsonDataTask(session) <| decoder) <| request
 }
 
 /// Fetches a GitHub user from an access token.
-func fetchGitHubUser(accessToken: GitHubAccessToken) -> EitherIO<Prelude.Unit, GitHubUser> {
+func fetchGitHubUser(accessToken: GitHubAccessToken) -> EitherIO<Error, GitHubUser> {
 
   let request = URLRequest(url: URL(string: "https://api.github.com/user")!)
     |> \.allHTTPHeaderFields .~ [
       "Authorization": "token \(accessToken.accessToken)",
       "Accept": "application/vnd.github.v3+json"
   ]
 
-  return .init(
-    run: .init { callback in
-      session
-        .dataTask(with: request) { data, response, error in
-          callback(
-            data.flatMap { try? JSONDecoder().decode(GitHubUser.self, from: $0) }
-              .map(Either.right)
-              ?? Either.left(unit)
-          )
-        }
-        .resume()
+  return (jsonDataTask(session) <| decoder) <| request
+}
+
+// TODO: Move to Prelude
+public extension Either where L == Error {
+  public static func wrap(_ f: @escaping () throws -> R) -> Either {
+    do {
+      return .right(try f())
+    } catch let error {
+      return .left(error)
+    }
+  }
+}
+
+// TODO: Move to PreludeFoundation?
+public func jsonDataTask<A>(_ session: URLSession)
```

Comment:
Some cleanup so that we're not dealing with some of the messier parts of `URLSession`.